### PR TITLE
Api exercises download without authentication

### LINKED ISSUE #401: [API] Download exercises by name without authentication
An unauthenticated user should be able to download an exercise by knowing its name.

### PR DESCRIPTION
**Pull request by makevertti**
closes #401 
Jenkins build: http://ohtu-jenkins.testmycode.io/job/tmc-server-api-exercises-download/13/

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -124,6 +124,11 @@ def respond_not_found(msg = 'Not Found')
     respond_with_error(msg, 404)
   end
 
+  def respond_not_found_with_auth_bypass(msg = 'Not Found')
+    authorize! :read, nil
```
Comment (nygrenh) [inline app/controllers/application_controller.rb]: `skip_authorization_check`

Comment (jackrl) [inline app/controllers/application_controller.rb]: Didn't we already agree that it doesn't work?

Comment (jackrl) [inline app/controllers/application_controller.rb]: We could maybe add a cancan definition for errors and use that instead

Comment (nygrenh) [inline app/controllers/application_controller.rb]: I remember that we concluded `skip_authorization_check` shouldn't be used  as a before action but I have no idea why it wouldn't work when called explicitly.

File: app/controllers/api/v8/exercises_controller.rb [hunk 2]
```diff
@@ -64,28 +64,59 @@ class Api::V8::ExercisesController < Api::V8::BaseController
...
-            solution_visible_after: ex.solution_visible_after,
-            deadline: ex.deadline_for(current_user),
-            disabled: ex.disabled?
-        }
-      end
-      authorize! :read, visible
-      present(presentable)
+    course = Course.find_by!(id: params[:id]) if (params[:id])
+    course ||= Course.find_by!(name: "#{params[:slug]}-#{params[:name]}")
+    exercises = Exercise.where(course_id: course.id)
```
Comment (nygrenh) [inline app/controllers/api/v8/exercises_controller.rb]: What happened to `.includes(:available_points)`?

Comment (jackrl) [inline app/controllers/api/v8/exercises_controller.rb]: Fixed. Must have been removed on rebase by mistake.

**Pull request merged**